File: frontend/src/services/intakes.ts
import { apiGet, apiPatch, apiPost } from "./api";

export type IntakeCreateRequest = {
  customer_name: string;
  environment_name: string;
  environment_type: "DEV" | "TEST" | "PROD";
};

export type IntakeCreateResponse = {
  intake_id: string;
  status: string;
  created_utc: number;
};

export type IntakeDetailResponse = {
  intake_id: string;
  status: string;
  customer_name: string;
  environment_name: string;
  environment_type: string;
  applications: unknown[];
};

export type IntakeUpdateRequest = {
  applications?: {
    application_name: string;
    product_line: string;
    current_version: string;
    target_version: string;
    modules_enabled: string[];
  }[];
  primary_technical_contact?: {
    name: string;
    email?: string | null;
  };
  primary_business_contact?: {
    name: string;
    email?: string | null;
  };
  environment_count?: number;
  environment_classification?: string[];
  upgrade_sequence?: string[];
  vendor_kb_documents?: {
    article_number: string;
    title: string;
    publication_date: string;
    source_link: string;
  }[];
};

export type IntakeValidateResponse = {
  intake_id: string;
  status: string;
  completeness_score: number;
  missing_required_fields: string[];
  warnings: string[];
};

export type StartAnalysisResponse = {
  analysis_id: string;
  status: string;
  started_utc: number;
};

export function createIntake(
  payload: IntakeCreateRequest,
): Promise<IntakeCreateResponse> {
  return apiPost<IntakeCreateResponse, IntakeCreateRequest>("/intakes", payload);
}

export function getIntake(id: string): Promise<IntakeDetailResponse> {
  return apiGet<IntakeDetailResponse>(`/intakes/${id}`);
}

export function updateIntake(id: string, payload: IntakeUpdateRequest) {
  return apiPatch(`/intakes/${id}`, payload);
}

export function validateIntake(id: string): Promise<IntakeValidateResponse> {
  return apiPost<IntakeValidateResponse, Record<string, never>>(
    `/intakes/${id}/validate`,
    {},
  );
}

export function startAnalysis(id: string): Promise<StartAnalysisResponse> {
  return apiPost<StartAnalysisResponse, { run_mode: string }>(
    `/intakes/${id}/analyses`,
    { run_mode: "STANDARD" },
  );
}

File: backend/app/api/v1/intakes.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.core.auth import AuthenticatedUser, UserRole, require_roles
from app.db.session import get_db
from app.schemas.intake_api import (
    IntakeCreateRequest,
    IntakeCreateResponse,
    IntakeDetailResponse,
    IntakeUpdateRequest,
    IntakeUpdateResponse,
    IntakeValidateResponse,
    StartAnalysisRequest,
    StartAnalysisResponse,
)
from app.services.intake_workflow_service import IntakeWorkflowService
from app.services.usage_event_service import UsageEventService

router = APIRouter(tags=["intakes"])
service = IntakeWorkflowService()
usage_event_service = UsageEventService()


@router.post("/intakes", response_model=IntakeCreateResponse)
def create_intake(
    payload: IntakeCreateRequest,
    db: Session = Depends(get_db),
    current_user: AuthenticatedUser = Depends(
        require_roles(UserRole.ANALYST, UserRole.ADMIN)
    ),
) -> IntakeCreateResponse:
    result = service.create_intake(db, payload)

    usage_event_service.record_event(
        db=db,
        event_type="INTAKE_CREATED",
        actor_role=current_user.role.value,
        actor_user_id=current_user.user_id,
        entity_type="INTAKE",
        entity_id=result.intake_id,
        event_payload={
            "environment_type": payload.environment_type.value,
            "customer_name": payload.customer_name,
            "environment_name": payload.environment_name,
        },
    )

    return result


@router.get("/intakes/{intake_id}", response_model=IntakeDetailResponse)
def get_intake(
    intake_id: str,
    db: Session = Depends(get_db),
    _: AuthenticatedUser = Depends(
        require_roles(
            UserRole.VIEWER,
            UserRole.ANALYST,
            UserRole.REVIEWER,
            UserRole.ADMIN,
        )
    ),
) -> IntakeDetailResponse:
    result = service.get_intake(db, intake_id)
    if not result:
        raise HTTPException(status_code=404, detail="Intake not found")
    return result


@router.patch("/intakes/{intake_id}", response_model=IntakeUpdateResponse)
def update_intake(
    intake_id: str,
    payload: IntakeUpdateRequest,
    db: Session = Depends(get_db),
    _: AuthenticatedUser = Depends(
        require_roles(UserRole.ANALYST, UserRole.ADMIN)
    ),
) -> IntakeUpdateResponse:
    result = service.update_intake(db, intake_id, payload)
    if not result:
        raise HTTPException(status_code=404, detail="Intake not found")
    return result


@router.post("/intakes/{intake_id}/validate", response_model=IntakeValidateResponse)
def validate_intake(
    intake_id: str,
    db: Session = Depends(get_db),
    current_user: AuthenticatedUser = Depends(
        require_roles(UserRole.ANALYST, UserRole.ADMIN)
    ),
) -> IntakeValidateResponse:
    result = service.validate_intake(db, intake_id)
    if not result:
        raise HTTPException(status_code=404, detail="Intake not found")

    event_type = "INTAKE_VALIDATED" if result.status == "INTAKE_VALIDATED" else "INTAKE_BLOCKED"

    usage_event_service.record_event(
        db=db,
        event_type=event_type,
        actor_role=current_user.role.value,
        actor_user_id=current_user.user_id,
        entity_type="INTAKE",
        entity_id=result.intake_id,
        event_payload={
            "status": result.status,
            "completeness_score": result.completeness_score,
            "missing_required_fields": result.missing_required_fields,
            "warnings": result.warnings,
        },
    )

    return result


@router.post("/intakes/{intake_id}/analyses", response_model=StartAnalysisResponse)
def start_analysis(
    intake_id: str,
    _: StartAnalysisRequest,
    db: Session = Depends(get_db),
    current_user: AuthenticatedUser = Depends(
        require_roles(UserRole.ANALYST, UserRole.ADMIN)
    ),
) -> StartAnalysisResponse:
    result = service.start_analysis(db, intake_id)
    if not result:
        raise HTTPException(
            status_code=400,
            detail="Intake must exist and be INTAKE_VALIDATED before analysis can start",
        )

    usage_event_service.record_event(
        db=db,
        event_type="ANALYSIS_STARTED",
        actor_role=current_user.role.value,
        actor_user_id=current_user.user_id,
        entity_type="ANALYSIS",
        entity_id=result.analysis_id,
        related_analysis_id=result.analysis_id,
        event_payload={
            "intake_id": intake_id,
            "status": result.status,
            "started_utc": result.started_utc,
        },
        commit=False,
    )

    if result.status != "ANALYSIS_RUNNING":
        usage_event_service.record_event(
            db=db,
            event_type="ANALYSIS_COMPLETED",
            actor_role=current_user.role.value,
            actor_user_id=current_user.user_id,
            entity_type="ANALYSIS",
            entity_id=result.analysis_id,
            related_analysis_id=result.analysis_id,
            event_payload={
                "intake_id": intake_id,
                "status": result.status,
                "started_utc": result.started_utc,
            },
            commit=True,
        )
    else:
        db.commit()

    return result